io.c

```diff
@@ -2244,6 +2237,166 @@ static int io_sqe_files_register(struct io_ring_ctx *ctx, void __user *arg,
+	if (arg.len & ~PAGE_MASK) {
+		pr_err("%s: mapped len is not aligned", __func__);
+		goto err;
+	}
+
+	if ((uintptr_t)arg.base & ~PAGE_MASK) {
+		pr_err("%s: mapped addr is not aligned", __func__);
+		goto err;
+	}
+
+	if (arg.max_len & ~PAGE_MASK) {
+		pr_err("%s: mapped max len is not aligned", __func__);
+		goto err;
+	}
+
+	if (arg.max_len < arg.len) {
+		pr_err("%s: max len less than len", __func__);
+		goto err;
+	}
+
+	for (buf_index = 0; buf_index < PXD_IO_MAX_USER_BUFS; ++buf_index) {
+		imu = &ctx->user_bufs[buf_index];
+		if (imu->ubuf <= (uintptr_t) arg.base &&
+		    (uintptr_t) arg.base + arg.len <= imu->ubuf + imu->len) {
+			offset = ((uintptr_t)arg.base - imu->ubuf) >> PAGE_SHIFT;
+			pr_info("found existing mapping %d, off %ld", buf_index, offset);
+			break;
+		}
+	}
+
+	if (buf_index == PXD_IO_MAX_USER_BUFS) {
+		existing = false;
+		for (buf_index = 0; buf_index < PXD_IO_MAX_USER_BUFS; ++buf_index) {
+			pr_info("new mapping %d, ", buf_index);
+			if (ctx->user_bufs[buf_index].ubuf == 0)
+				break;
+		}
+	}
+
+	if (buf_index == PXD_IO_MAX_USER_BUFS) {
+		pr_err("%s: all busy", __func__);
+		return -EBUSY;
+	}
+
+	imu = &ctx->user_bufs[buf_index];
+
+	/*
+         * Don't impose further limits on the size and buffer
+         * constraints here, we'll -EINVAL later when IO is
+         * submitted if they are wrong.
+         */
+	ret = -EFAULT;
+	if (!arg.base || !arg.len)
+		goto err;
+
+	/* arbitrary limit, but we need something */
+	if (arg.len > SZ_4G)
+		goto err;
+
+	ubuf = (unsigned long) arg.base;
+	end = (ubuf + arg.len) >> PAGE_SHIFT;
+	start = ubuf >> PAGE_SHIFT;
+	nr_pages = end - start;
+	max_nr_pages = arg.max_len >> PAGE_SHIFT;
+
+	ret = -ENOMEM;
+	pages = kvmalloc_array(nr_pages, sizeof(struct page *),
+		GFP_KERNEL);
+	vmas = kvmalloc_array(nr_pages, sizeof(struct vm_area_struct *),
+		GFP_KERNEL);
+	if (!pages || !vmas) {
+		goto err;
+	}
+
+	imu->bvec = kvcalloc(max_nr_pages, sizeof(struct bio_vec),
+		GFP_KERNEL);
+	if (!imu->bvec) {
+		goto err;
+	}
+
+	ret = 0;
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 8, 0)
+	down_read(&current->mm->mmap_sem);
+#else
+	down_read(&current->mm->mmap_lock);
+#endif
+	pret = get_user_pages(ubuf, nr_pages, FOLL_WRITE, pages, vmas);
+	pr_info("%s: %ld bytes %d pages", __func__, arg.len, nr_pages);
+	if (pret < 0) {
+		ret = pret;
+	} else if (pret != nr_pages) {
+		ret = -EFAULT;
+	}
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 8, 0)
+	up_read(&current->mm->mmap_sem);
+#else
+	up_read(&current->mm->mmap_lock);
+#endif
+	if (ret) {
+		/*
+                 * if we did partial map release any pages we did get
+                 */
+		if (pret > 0) {
+			for (j = 0; j < pret; j++)
+				put_page(pages[j]);
+		}
+		kvfree(imu->bvec);
+		imu->bvec = NULL;
+		goto err;
+	}
+
+	off = ubuf & ~PAGE_MASK;
+	size = arg.len;
+	for (j = 0; j < nr_pages; j++) {
+		size_t vec_len;
+
+		vec_len = min_t(size_t, size, PAGE_SIZE - off);
+		imu->bvec[j + offset].bv_page = pages[j];
```

Comment:
if this is a previously mapped range, then previous mapping is overwritten here for the new range passed. This would leave previous page pinned in memory and references lost to it.

Also uring allows only initializing one set of fixed buffers. This currently extends to allow 16 such sets.

Reply:
added check

it allows for multiple set of buffers just the same, only there is a requirement to be done in one shot. This implementation relaxes that requirement.